gsplat/cuda/csrc/relocation.cu

```diff
@@ -0,0 +1,67 @@
+#include "bindings.h"
+
+// Equation (9) in "3D Gaussian Splatting as Markov Chain Monte Carlo"
+__global__ void compute_relocation_kernel(
+    int P,
+    float *old_opacities,
+    float *old_scales,
+    int *ratios,
+    float *binoms,
+    int n_max,
+    float *new_opacities,
+    float *new_scales
+) {
+    int idx = threadIdx.x + blockIdx.x * blockDim.x;
+    if (idx >= P)
+        return;
+
+    int n_idx = ratios[idx];
+    float denom_sum = 0.0f;
+
+    // compute new opacity
+    new_opacities[idx] = 1.0f - powf(1.0f - old_opacities[idx], 1.0f / n_idx);
+
+    // compute new scale
+    for (int i = 1; i <= n_idx; ++i) {
+        for (int k = 0; k <= (i - 1); ++k) {
+            float bin_coeff = binoms[(i - 1) * n_max + k];
+            float term =
+                (pow(-1, k) / sqrt(k + 1)) * pow(new_opacities[idx], k + 1);
+            denom_sum += (bin_coeff * term);
+        }
+    }
+    float coeff = (old_opacities[idx] / denom_sum);
+    for (int i = 0; i < 3; ++i)
+        new_scales[idx * 3 + i] = coeff * old_scales[idx * 3 + i];
+}
+
+std::tuple<torch::Tensor, torch::Tensor> compute_relocation_tensor(
+    torch::Tensor &old_opacities,
+    torch::Tensor &old_scales,
+    torch::Tensor &ratios,
+    torch::Tensor &binoms,
+    const int n_max
+) {
```

Comment:
Adding `DEVICE_GUARD` is necessary when it is being used in a torch distributed training system. Also would be nice to apply `CHECK_INPUT` to the input tensors to make sure they are contiguous in memory, otherwise it could silently lead to incorrect results that is hard to debug. See reference:

https://github.com/nerfstudio-project/gsplat/blob/0ac847d6efde31dba8c289c67241321a3cf9821f/gsplat/cuda/csrc/projection.cu#L159-L161